src/motion/features/viewport/observers.ts

```diff
@@ -0,0 +1,71 @@
+type IntersectionHandler = (entry: IntersectionObserverEntry) => void
+
+interface ElementIntersectionObservers {
+    [key: string]: IntersectionObserver
+}
+
+/**
+ * Map an IntersectionHandler callback to an element. We only ever make one handler for one
+ * element, so even though these handlers might all be triggered by different
+ * observers, we can keep them in the same map.
+ */
+const observerCallbacks = new WeakMap<Element, IntersectionHandler>()
+
+/**
+ * Multiple observers can be created for multiple element/document roots. Each with
+ * different settings. So here we store dictionaries of observers to each root,
+ * using serialised settings (threshold/margin) as lookup keys.
+ */
+const observers = new WeakMap<
+    Element | Document,
+    ElementIntersectionObservers
+>()
+
+const fireObserverCallback = (entry: IntersectionObserverEntry) => {
+    observerCallbacks.get(entry.target)?.(entry)
+}
+
+const fireAllObserverCallbacks: IntersectionObserverCallback = (entries) => {
+    entries.forEach(fireObserverCallback)
+}
+
+function initIntersectionObserver({
+    root,
+    ...options
+}: IntersectionObserverInit): IntersectionObserver {
+    /**
+     * If we don't have an observer lookup map for this root, create one.
+     */
+    if (!observers.has(root || document)) {
+        observers.set(root || document, {})
+    }
+    const rootObservers = observers.get(root || document)!
+
+    const key = JSON.stringify(options)
+
+    /**
+     * If we don't have an observer for this combination of root and settings,
+     * create one.
+     */
+    if (!rootObservers[key]) {
```

Comment:
Because the root is not part of the key, we need to check that the root is actually the same as the existing intersectionObserver, like we do [here](https://github.com/framer/FramerStudio/blob/d6077be8587602ac1a2e8eb901da57a91aaffc30/src/library/src/modules/useAppearEffect.ts#L63)

Reply:
We've already selected from a subset of intersection observers for this root

Reply:
💯